### PR TITLE
Fixing constructors from 0-element arrays

### LINKED ISSUE #557: error creating empty MVector
This seems like it should work:
```jl
julia> MVector(SVector{0,Float64}())
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] MArray{Tuple{0},T,1,0} where T(::Tuple{}) at /Users/stevenj/.julia/packages/StaticArrays/3ENSR/src/MVector.jl:20
 [2] Type at /Users/stevenj/.julia/packages/StaticArrays/3ENSR/src/MVector.jl:19 [inlined]
 [3] MArray{Tuple{S},T,1,S} where T where S(::SArray{Tuple{0},Float64,1,0}) at /Users/stevenj/.julia/packages/StaticArrays/3ENSR/src/convert.jl:5
 [4] top-level scope at none:0
```

### PR DESCRIPTION
**Pull request by mateuszbaran**
Maybe not the prettiest solution but still fixes #557 and #782.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
That failure looks like a Julia 1.0 bug and I don't see an obvious way to work around it 😕 .

**Comment by mateuszbaran** [comment]
Apart from not solving the problem of converting empty `StaticArray`s to `SizedArray` or `SizedMatrix` on Julia 1.0, this seems to work now :slightly_smiling_face: . Anyway, I think these two cases can be fixed in another PR if someone considers it worth fixing.

**Review by c42f — APPROVED** [review]
Ack, sorry for the delay in replying to this. I'm a little unsure about this solution in the sense that it requires a new constructor for each StaticArray subtype. I feel like that shouldn't be necessary...

However, I also know that the generic dispatch for StaticArray constructors is quite hairy and hard to modify and that this does fix the bugs. So I think on balance we should just go with it.

(Regarding 1.0 - at this point it's basically EOL and I'm looking forward to removing support for it once 1.6 (or 1.7?) is declared the new LTS. So I'm not too worried about that.)

**Comment by mateuszbaran** [comment]
> I'm a little unsure about this solution in the sense that it requires a new constructor for each StaticArray subtype. I feel like that shouldn't be necessary...

After working with this a bit I feel that a part of hairiness of static array constructors is trying to define things for all subtypes of `StaticArray`. But I may be wrong. At the very least not all subtypes of `StaticArray` are constructible just from elements.

File: src/MArray.jl [hunk 1]
```diff
@@ -71,7 +71,7 @@ end
     end
 end
 
-@inline MArray(a::StaticArray) = MArray{size_tuple(Size(a))}(Tuple(a))
+@inline MArray(a::StaticArray{S,T}) where {S<:Tuple,T} = MArray{size_tuple(Size(a)),T}(Tuple(a))
```
Comment (mateuszbaran) [inline src/MArray.jl:74]:
```suggestion
@inline MArray(a::StaticArray{S,T}) where {S<:Tuple,T} = MArray{S,T}(Tuple(a))
```

**Pull request merged**